src/control/system/database.go

```diff
@@ -874,3 +876,38 @@ func (db *Database) UpdatePoolService(ps *PoolService) error {
 
 	return nil
 }
+
+// OnEvent handles events and updates system database accordingly.
+func (db *Database) OnEvent(_ context.Context, evt *events.RASEvent) {
+	switch evt.ID {
+	case events.RASPoolSvcReplicasUpdate:
+		ei := evt.GetPoolSvcInfo()
+		if ei == nil {
+			db.log.Error("no extended info in PoolSvcReplicasUpdate event received")
+			return
+		}
+		db.log.Debugf("processing RAS event %q for pool %s with info %+v on host %q",
+			evt.Msg, evt.PUUID, ei, evt.Hostname)
+
+		uuid, err := uuid.Parse(evt.PUUID)
+		if err != nil {
+			db.log.Errorf("failed to parse pool UUID %q: %s", evt.PUUID, err)
+			return
+		}
+
+		ps, err := db.FindPoolServiceByUUID(uuid)
+		if err != nil {
+			db.log.Errorf("failed to find pool with UUID %q: %s", evt.PUUID, err)
+			return
+		}
+
+		db.log.Debugf("update pool %s (state=%s) svc ranks %v->%v",
+			ps.PoolUUID, ps.State, ps.Replicas, ei.SvcReplicas)
+
+		ps.Replicas = RanksFromUint32(ei.SvcReplicas)
+
+		if err := db.UpdatePoolService(ps); err != nil {
+			db.log.Errorf("failed to apply pool service update: %s", err)
+		}
```

Comment:
I think this is fine for now, but we need to take another look at some of the ways we're accessing the system database. There is a potential race here where two events could come in quickly with conflicting service replicas. We're doing a read lock to get the pool service entry, and then UpdatePoolService() results in the update being committed via raft, but in the interim there's no serialization to ensure that PoolServiceReplicasUpdate events are handled sequentially. I'll have to think about this some more. We may be able to leverage the leader term from the incoming payload to ensure that we're only ever going forward.